File: backend-fastapi/lib/types_/filters.py
from typing import Any, Literal, TypedDict

# ---- Public Types for searching data ----

FilterOperation = Literal[
    "eq",
    "ne",
    "null",
    "in",
    "nin",
    "lt",
    "lte",
    "gt",
    "gte",
    "like",
    "ilike",
]


# Using TypedDict instead of dataclass because of easy serialization
class Filter(TypedDict):
    op: FilterOperation
    val: Any


type WhereFilters[Searchables: str] = dict[Searchables, list[Filter]]

File: backend-express/src/lib/types/filters.ts
export type FilterOperation =
    | "eq"
    | "ne"
    | "null"
    | "in"
    | "nin"
    | "lt"
    | "lte"
    | "gt"
    | "gte"
    | "like"
    | "ilike";

export type Filter = { op: FilterOperation; val: any };

export type WhereFilters<T extends string> = {
    [K in T]?: Filter[];
};
